File: algo/103. Binary Tree Zigzag Level Order Traversal.js
// BFS

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if (root == null) {
        return [];
    }
    
    var res = [];
    var queue = [root];
    var levelIndex = 1;
    while (queue.length != 0) {
        var levelRes = [];
        var levelSize = queue.length;

        for (var i=0; i<levelSize; i++) {
            let currNode = queue.shift();
            
            // use stack for odd line, use queue for even line
            if (levelIndex % 2 == 1) {
                levelRes.push(currNode.val);
            } else {
                levelRes.unshift(currNode.val);
            }            
            if (currNode.left) {
                queue.push(currNode.left);
            }

            if (currNode.right) {
                queue.push(currNode.right);
            }
        }
        
        levelIndex ++;
        res.push(levelRes);
    }
    
    return res;
};

File: round2/103. Binary Tree Zigzag Level Order Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        List<IList<int>> res = new List<IList<int>>();
        if (root == null) return res;
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        bool isOddLevel = true;
        while(queue.Count != 0) {
            var levelRes = new List<int>();
            int levelSize = queue.Count;
            for (int i=0; i<levelSize; i++) {
                TreeNode curr = queue.Dequeue();
                levelRes.Add(curr.val);
                if (curr.left != null) queue.Enqueue(curr.left);
                if (curr.right != null) queue.Enqueue(curr.right);
            }
            if (!isOddLevel) levelRes.Reverse();
            res.Add(levelRes);
            isOddLevel = !isOddLevel;
        }
        return res;
    }
}